### PR TITLE
feat(storage): improve error messages on stalled uploads

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9744)

### CODE REVIEW

**Review by scotthart — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9744)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @coryan)

File: google/cloud/storage/internal/retry_client.cc [hunk 1]
```diff
@@ -154,6 +154,42 @@ bool UploadChunkOnFailure(RetryPolicy& retry_policy, int& count,
   return retry_policy.OnFailure(status);
 }
 
+Status RetryError(Status const& last_status, RetryPolicy const& retry_policy,
+                  char const* error_message) {
```
Comment (devbww) [inline google/cloud/storage/internal/retry_client.cc:158]: `error_message` seems a little misleading in the context of `"... in <error_message>"`.

Previously we seem to have used `char const* location` or `char const* function_name`.

File: google/cloud/storage/internal/retry_client.cc [hunk 2]
```diff
@@ -154,6 +154,42 @@ bool UploadChunkOnFailure(RetryPolicy& retry_policy, int& count,
...
+
+Status PartialWriteStatus(int error_count, int upload_count,
+                          std::int64_t committed_size,
+                          std::int64_t expected_committed_size,
+                          Status last_status) {
+  if (error_count > 0) return last_status;
+  std::ostringstream os;
+  os << "All requests (" << upload_count << ") have succeeded, but they have"
+     << " not completed the full write. The expected committed size is "
+     << expected_committed_size << " the current committed size is "
```
Comment (devbww) [inline google/cloud/storage/internal/retry_client.cc:188]: s/the/and the/?

File: google/cloud/storage/internal/retry_client.cc [hunk 3]
```diff
@@ -494,8 +530,10 @@ StatusOr<EmptyResponse> RetryClient::DeleteResumableUpload(
 //
 StatusOr<QueryResumableUploadResponse> RetryClient::UploadChunk(
     UploadChunkRequest const& request) {
-  Status last_status(StatusCode::kDeadlineExceeded,
-                     "Retry policy exhausted before first attempt was made.");
+  auto const initial_status =
```
Comment (devbww) [inline google/cloud/storage/internal/retry_client.cc:533]: `initial_status` seems to be unused.

Comment (coryan) [reply google/cloud/storage/internal/retry_client.cc]: Fixed.

File: google/cloud/storage/internal/retry_client.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/retry_client.cc]: Fixed

File: google/cloud/storage/internal/retry_client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/retry_client.cc]: Fixed

**Pull request merged**